### PR TITLE
feat(storage): auto double part size when body exceeds 10k parts

### PR DESCRIPTION
**Pull request by AllanZhengYP**
#### Description of changes

The multipart upload provider used to use 5 MB fixed chunk size. Given the 10k parts limit of S3 service, the provider can upload 50GB object at most. With this change, the provider will double the chunk size(10 MB, 20 MB etc) until the object can be chunked within 10k parts, or object exceeds the S3 limit(5 TB). [The Swift SDK](https://github.com/aws-amplify/amplify-swift/blob/7f094a9e88612b5b8fda1bce01ce71e7f1915053/AmplifyPlugins/Storage/Sources/AWSS3StoragePlugin/Support/Internal/StorageUploadPart.swift#L200) uses the same strategy. 

This change also refactors unit test and fixed a bug where clean up method would be called in single part upload.

#### Issue #, if available

Resolves: https://github.com/aws-amplify/amplify-js/issues/8772

#### Description of how you validated changes
* [Unit and Integ test](https://app.circleci.com/pipelines/github/aws-amplify/amplify-js/17322/workflows/54e16700-29cc-4aa6-8e05-053efc56ef30)

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [ ] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by haverchuck — APPROVED** [review]

**Review by haverchuck — DISMISSED** [review]
LGTM

File: packages/storage/src/common/StorageConstants.ts [hunk 1]
```diff
@@ -1,10 +1,20 @@
...
 export const localTestingStorageEndpoint = 'http://localhost:20005';
 
 export const UPLOADS_STORAGE_KEY = '__uploadInProgress';
+
+const MB = 1024 * 1024;
+const GB = 1024 * MB;
+const TB = 1024 * GB;
+
+export const DEFAULT_PART_SIZE = 5 * MB;
+export const MAX_OBJECT_SIZE = 5 * TB;
```
Comment (haverchuck) [inline packages/storage/src/common/StorageConstants.ts:18]: What imposes this limitation?

**Review by haverchuck — APPROVED** [review]

**Review by jimblanc — APPROVED** [review]

File: packages/storage/__tests__/providers/AWSS3ProviderManagedUpload-unit-test.ts [hunk 2]
```diff
@@ -1,371 +1,491 @@
...
+			} catch (error) {
+				expect(error.message).toEqual(
+					expect.stringContaining(
+						'File size bigger than S3 Object limit of 5TB'
+					)
+				);
+			}
+		});
+
+		test('error case: upload a body that splits in two parts but second part fails', async () => {
```
Comment (stocaaro) [inline packages/storage/__tests__/providers/AWSS3ProviderManagedUpload-unit-test.ts:312]: nit: Have mixed feelings about this test. Its super long, which makes it hard to follow, however its doing a complicated mock setup to test a complicated multi-step interaction. Long term, I might think about restructuring this to factor out some of the repetition / write helpers to tell the story more succinctly.

I really appreciate that there isn't a ton of hidden detail elsewhere in this file that makes this work.

Vote we commit it as is, expressing an opinion that isn't currently reflected in our codebase that I would like to see us refactor towards.

**Review by stocaaro — APPROVED** [review]
A couple nits, but nothing that needs changing in this PR. 🚢 it!

File: packages/storage/src/providers/AWSS3UploadTask.ts [hunk 3]
```diff
@@ -227,11 +226,9 @@ export class AWSS3UploadTask implements UploadTask {
 	}
 
 	private _validateParams() {
-		if (this.file.size / this.partSize > MAX_PARTS) {
+		if (this.totalBytes > MAX_OBJECT_SIZE) {
 			throw new Error(
-				`Too many parts. Number of parts is ${
-					this.file.size / this.partSize
-				}, maximum is ${MAX_PARTS}.`
+				`File size bigger than S3 Object limit of 5TB, got ${this.totalBytes} Bytes`
```
[the remark this answers is not in the document]
Comment (stocaaro) [reply packages/storage/src/providers/AWSS3UploadTask.ts:231]: Excellent question. Not sure we've held that standard historically.  We should look at having stronger API contracts moving forward, maybe clearly separated so its easier to review those parts more deliberately.

File: packages/storage/__tests__/providers/AWSS3ProviderManagedUpload-unit-test.ts (hunk 2 above)
Comment (AllanZhengYP) [reply packages/storage/__tests__/providers/AWSS3ProviderManagedUpload-unit-test.ts:312]: I tend to agree. I do have more idea of refactoring the uploaders, especially the hundreds lines of duplicated implementations in `S3UploadTask` and `S3ManagedUploader` class. Btw, the diff was busted for some reason, this test case was not really in diff.

File: packages/storage/src/providers/AWSS3UploadTask.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (AllanZhengYP) [reply packages/storage/src/providers/AWSS3UploadTask.ts:231]: Just for the context, the error is changed here because with the new implementation, it's not possible to reach the part number limit. S3 limit for each part size is 5 GB, for any object is 5 TB(1000 parts). So if we increase the part size on demand, the customer will reach the 5 TB limit first instead of the parts number.
